external-crates/move/crates/move-compiler/src/linters/self_assignment.rs

```diff
@@ -0,0 +1,145 @@
+            match (&lhs.exp.value, &inner_exp.exp.value) {
+                (
+                    UnannotatedExp_::Borrow(_, inner_lhs, lhs_field),
+                    UnannotatedExp_::Borrow(_, inner_rhs, rhs_field),
+                ) if inner_lhs == inner_rhs && lhs_field == rhs_field => {
+                    self.report_self_assignment(loc);
+                    true
+                }
+                (
+                    UnannotatedExp_::Copy {
+                        var: sp!(_, lhs), ..
+                    },
+                    UnannotatedExp_::Copy {
+                        var: sp!(_, rhs), ..
+                    },
+                ) if lhs == rhs => {
+                    self.report_self_assignment(loc);
+                    true
+                }
+                _ => false,
+            }
+        } else {
+            false
+        }
+    }
+
+    fn check_assign(&mut self, value_list: &LValueList, assign_exp: &T::Exp, loc: Loc) -> bool {
+        if let UnannotatedExp_::ExpList(rhs_expressions) = &assign_exp.exp.value {
+            let is_self_assignment =
+                value_list
+                    .value
+                    .iter()
+                    .zip(rhs_expressions.iter())
+                    .all(|(lhs_value, rhs_exp)| {
+                        if let (
+                            sp!(
+                                _,
+                                LValue_::Var {
+                                    var: sp!(_, lhs),
+                                    ..
+                                }
+                            ),
+                            Single(inner_exp, _),
+                        ) = (lhs_value, rhs_exp)
+                        {
+                            matches!(
+                                &inner_exp.exp.value,
+                                UnannotatedExp_::Copy {
+                                    var: sp!(_, rhs),
+                                    ..
+                                }
+                                | UnannotatedExp_::Move {
+                                    var: sp!(_, rhs),
+                                    ..
+                                }
+                                if lhs == rhs
+                            )
+                        } else {
+                            false
+                        }
+                    });
+
+            if is_self_assignment && !value_list.value.is_empty() {
+                self.report_self_assignment(loc);
+                true
+            } else {
+                false
+            }
+        } else {
+            false
+        }
+    }
+
+    fn report_self_assignment(&mut self, loc: Loc) {
+        let msg = "Explicit self-assignment detected for variable. Consider removing it to clarify intent.".to_string();
+        let diag = diag!(SELF_ASSIGNMENT_DIAG, (loc, msg));
+        self.env.add_diag(diag);
+    }
+}
```

Comment:
The function takes three parameters:
`&mut self:`A mutable reference to the struct this method belongs to.
`value_list:`A reference to an LValueList, which likely contains the left-hand side of the assignment.
`assign_exp`: A reference to the right-hand side of the assignment.
`loc`: A Loc value, probably representing the location of this assignment in the source code.

It first checks if the `assign_exp` is an `ExpList` (a list of expressions).
If it is, it then checks for self-assignment by iterating over the left-hand side values (`value_list.value`) and the right-hand side expressions (`rhs_expressions`) simultaneously.
For each pair, it checks if:

The left-hand side is a `LValue_::Var`.
The right-hand side is a `Single` expression.
The right-hand side expression is either a `Copy` or `Move` operation.
The variable being copied or moved on the right side is the same as the variable on the left side.

If all pairs in the assignment satisfy these conditions, `is_self_assignment` is set to true.